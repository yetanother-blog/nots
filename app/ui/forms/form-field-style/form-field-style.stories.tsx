import type { Meta, StoryObj } from '@storybook/react';
import { FormFieldStyle } from './form-field-style';

/**
 * ### Usage
 *
 * You can use the `FormFieldStyle` component to style form fields and add some
 * `children` to it. Or you can also use the `as` prop to render a different
 * element.
 *
 * ### `children`
 *
 * Basically, same code as demonstrated by the above example.
 *
 * ```jsx
 * <FormFieldStyle className="block h-11">
 *   <input
 *     className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
 *     id="name"
 *     type="text"
 *   />
 * </FormFieldStyle>;
 * ```
 */
const meta: Meta<typeof FormFieldStyle> = {
  title: 'Forms/FormFieldStyle',
  component: FormFieldStyle,
};

export default meta;
type Story = StoryObj<typeof FormFieldStyle>;

/**
 * The `FormFieldStyle` component is used to style form fields. It is used as a
 * wrapper around the `input` element.
 */
export const FormFieldStyleStory: Story = {
  parameters: {
    docs: {
      canvas: {
        sourceState: 'shown',
      },
    },
  },
  render: () => (
    <FormFieldStyle className="block h-11">
      <input
        className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
        id="name"
        type="text"
      />
    </FormFieldStyle>
  ),
};

export const AsInputDisabled: Story = {
  parameters: {
    docs: {
      canvas: {
        sourceState: 'shown',
      },
    },
  },
  render: () => (
    <FormFieldStyle className="block h-11" disabled>
      <input
        className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
        id="name"
        type="text"
      />
    </FormFieldStyle>
  ),
};

export const AsTextArea: Story = {
  render: () => (
    <FormFieldStyle
      className="block h-64 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      as="textarea"
    />
  ),
};

export const AsTextAreaDisabled: Story = {
  render: () => (
    <FormFieldStyle
      className="block h-64 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      as="textarea"
      disabled
    />
  ),
};

export const AsSelect: Story = {
  render: () => (
    <FormFieldStyle
      className="block h-11 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      as="select"
    >
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
    </FormFieldStyle>
  ),
};

export const AsSelectDisabled: Story = {
  render: () => (
    <FormFieldStyle
      className="block h-11 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      as="select"
      disabled
    >
      <option value="1">Option 1</option>
      <option value="2">Option 2</option>
      <option value="3">Option 3</option>
    </FormFieldStyle>
  ),
};
