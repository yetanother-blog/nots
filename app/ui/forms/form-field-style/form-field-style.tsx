import { clsx } from 'clsx';
import type { ElementType, ReactElement } from 'react';
import { forwardRef } from 'react';
import type { PolymorphicComponentPropWithRef } from '~/ui';

export type FormFieldStyleProps<C extends ElementType> =
  PolymorphicComponentPropWithRef<
    C,
    {
      isError?: boolean;
      disabled?: boolean;
      className?: string;
    }
  >;

export type FormFieldStyleComponent = <C extends ElementType = 'span'>(
  props: FormFieldStyleProps<C>
) => ReactElement | null;

// eslint-disable-next-line react/display-name
export const FormFieldStyle = forwardRef(
  <C extends ElementType = 'span'>(
    {
      as,
      children,
      isError,
      disabled,
      className,
      ...props
    }: FormFieldStyleProps<C>,
    ref?: unknown
  ) => {
    const Component = as || 'span';
    const disabledAttribute = as ? { disabled } : undefined;

    return (
      <Component
        ref={ref}
        className={clsx(
          'border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4',
          isError
            ? 'border-nots-danger-500  ring-nots-danger-300 focus-within:border-nots-danger-500  hover:border-nots-danger-500'
            : 'border-nots-grey-300  ring-nots-grey-100 focus-within:border-nots-grey-500  hover:border-nots-grey-500 dark:border-nots-grey-600  dark:ring-nots-grey-600 dark:focus-within:border-nots-grey-700  dark:hover:border-nots-grey-700',
          disabled && 'pointer-events-none bg-ws-grey-50',
          className
        )}
        {...disabledAttribute}
        {...props}
      >
        {children}
      </Component>
    );
  }
) as FormFieldStyleComponent;
