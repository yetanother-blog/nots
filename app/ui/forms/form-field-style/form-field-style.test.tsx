import { render } from '@testing-library/react';
import { FormFieldStyle } from './form-field-style';

describe('FormFieldStyle', () => {
  it('should render basic text input with style 😎', () => {
    const { container } = render(
      <FormFieldStyle className="block h-11">
        <input
          className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
          id="name"
          type="text"
        />
      </FormFieldStyle>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render basic text input with disabled style 🫣', () => {
    const { container } = render(
      <FormFieldStyle className="block h-11" disabled>
        <input
          className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
          id="name"
          type="text"
        />
      </FormFieldStyle>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render basic text input with error style 🚨', () => {
    const { container } = render(
      <FormFieldStyle className="block h-11" isError={true}>
        <input
          className="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
          id="name"
          type="text"
        />
      </FormFieldStyle>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render form field style as textarea 🦹‍♀️', () => {
    const { container } = render(
      <FormFieldStyle
        className="block h-64 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
        as="textarea"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should render form field style as textarea w/ disabled style 🫣', () => {
    const { container } = render(
      <FormFieldStyle
        className="block h-64 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
        as="textarea"
        disabled
      />
    );
    expect(container).toMatchSnapshot();
  });
});
