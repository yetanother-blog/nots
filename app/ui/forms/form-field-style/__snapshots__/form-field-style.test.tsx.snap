// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormFieldStyle > should render basic text input with disabled style 🫣 1`] = `
<div>
  <span
    class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 pointer-events-none bg-ws-grey-50 block h-11"
  >
    <input
      class="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      id="name"
      type="text"
    />
  </span>
</div>
`;

exports[`FormFieldStyle > should render basic text input with error style 🚨 1`] = `
<div>
  <span
    class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-danger-500  ring-ws-danger-300 focus-within:border-ws-danger-500  hover:border-ws-danger-500 block h-11"
  >
    <input
      class="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      id="name"
      type="text"
    />
  </span>
</div>
`;

exports[`FormFieldStyle > should render basic text input with style 😎 1`] = `
<div>
  <span
    class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 block h-11"
  >
    <input
      class="h-full w-full bg-transparent px-3 text-sm font-light text-ws-grey-900  placeholder-ws-grey-400  outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
      id="name"
      type="text"
    />
  </span>
</div>
`;

exports[`FormFieldStyle > should render form field style as textarea w/ disabled style 🫣 1`] = `
<div>
  <textarea
    class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 pointer-events-none bg-ws-grey-50 block h-64 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
    disabled=""
  />
</div>
`;

exports[`FormFieldStyle > should render form field style as textarea 🦹‍♀️ 1`] = `
<div>
  <textarea
    class="rounded-lg border transition-all focus-within:outline-none focus-within:ring-4 active:ring-4 border-ws-grey-300  ring-ws-primary-300 focus-within:border-ws-primary-500  hover:border-ws-primary-500 block h-64 w-full bg-transparent p-3 text-sm  font-light text-ws-grey-900 placeholder-ws-grey-400 outline-none read-only:text-ws-grey-600 disabled:text-ws-grey-400"
  />
</div>
`;
